refactor(core): implement flag to ignore ngZone in message event listener

AleksanderBodurri opened the pull request:
This flag allows message event listeners to prevent callbacks from executing within the NgZone if they contain a special `__ignore_ng_zone__` flag.

This flag is built with Angular DevTools in mind, where it prevents an infinite change detection loop in inspected applications that have message event listeners

Cycle we want to avoid by using this flag:
CD -> Inspected app emits componentTreeDirty event to DevTools -> DevTools emits event to get new component Tree from Inspected app -> Inspected app message event listener fires -> CD


closes #45572

AleksanderBodurri committed:
refactor(core): implement __ignore_ng_zone__ flag

This flag allows message event listeners to prevent callbacks from executing within the NgZone if they contain a special `__ignore_ng_zone__` flag.

This functionality is built with Angular DevTools in mind, where it prevents an infinite change detection loop in inspecting applications that have message event listeners:

CD -> Inspected app emits componentTreeDirty event to DevTools -> DevTools emits event to get new component Tree from Inspected app -> Inspected app message event listener fires -> CD

dgp1130 commented on devtools/src/app/app.component.ts:
```diff
@@ -15,5 +15,10 @@ import {Router} from '@angular/router';
+    window.addEventListener('message', (event) => {
+      this.messagesRecieved = this.messagesRecieved + 1;
+    });
+  }
```
Was this intended for testing? Can we delete this part of the change?

dgp1130 reviewed (changes requested):
Are there any tests in Zone.js or DevTools we should update?

It would be very cool if we could have an e2e test for DevTools which runs on an app that sends a `MessageEvent` and confirms that things don't break. That might be a very involved test though, so I don't think it's strictly blocking here. Just some kind of automated validation would be useful.

JiaLiPassion commented on packages/core/src/zone/ng_zone.ts:
```diff
@@ -419,6 +419,10 @@ function forkInnerZoneWithAngularBehavior(zone: NgZonePrivate) {
          applyArgs: any): any => {
+          if (shouldBeIgnoredByZone(applyArgs)) {
+            return delegate.invokeTask(target, task, applyThis, applyArgs);
+          }
```
I believe it is ok that we only add this logic to `onInvokeTask`, since `message` is an `eventTask` and triggers the `onInvokeTask` callback.

And `onInvoke` hook is triggered when `ngZone.run(fn)` is invoked.

dgp1130 replied:
Ok makes sense. An opt-out for `ngZone.run` doesn't make a whole lot of sense, so limiting to `onInvokeTask` sounds right.

JiaLiPassion reviewed (approved):
LGTM! Thank you.

mikezks commented:
@brandonroberts, @markostanimirovic 

Will this fix the issue with the Redux DevTools for NgRx in combination with the Angular DevTools as well?

brandonroberts commented:
> @brandonroberts, @markostanimirovic
> 
> Will this fix the issue with the Redux DevTools for NgRx in combination with the Angular DevTools as well?

Maybe. We recently introduced a change in NgRx 16.2 that lets you opt-out with StoreDevtools. https://github.com/ngrx/platform/issues/3839

It may be possible that we send this flag along with the message to the DevTools, but then it may show up in the action log. Needs more investigation there.

AleksanderBodurri committed:
fix(devtools): use the __ignore_ng_zone__ flag in devtools message bus' to prevent CD loop

Updates ChromeMessageBus, SamePageMessageBus and IframeMessageBus to use this new flag in their emit methods.

dgp1130 reviewed (approved):
LGTM, I'll leave the question of how to handle `__ignore_ng_zone__` check for callbacks with multiple arguments to you. I don't feel very strongly about it, just something which stood out to me.

dgp1130 commented on packages/core/src/zone/ng_zone.ts:
```diff
@@ -585,3 +589,19 @@ export function isStableFactory() {
+
+  if (applyArgs.length > 1) {
+    return false;
+  }
```
Followed up outside GitHub, apparently `task.invoke` is only ever called with a single argument:
https://github.com/angular/angular/blob/3658e25f2eb884fc0b1cea0203a5244d7c592581/packages/zone.js/lib/common/events.ts#L125

The ideal behavior in this case is a bit unclear, since we're not too sure what the semantics or expected contract would be if multiple arguments are passed in the future. For now we'll just restrict this to single argument calls, if it becomes a problem in the future we should hopefully have a better sense of what the right behavior is here.

AndrewKushnir commented:
[Presubmit](http://test/OCL:562076119:BASE:562076137:1693611269241:a39dd0de).

AndrewKushnir commented:
**Caretaker note**: TGP is "green".

AndrewKushnir commented:
@AleksanderBodurri could you please update commit messages to change them from `feat` to `refactor`? This would avoid a confusion in changelog, since this is not a feature that we want to expose publicly, it's more a compatibility improvement between Angular and DevTools.

AndrewKushnir commented:
**Caretaker note**: there was a rebase to update commit messages, but there were no changes to the code, so we can reuse a previous TGP which was "green".

jessicajaniuk commented:
This PR was merged into the repository by commit 4b54947c97568352decf4b54938d5a1eb5ae26f5.
